# Homework 3 (Woosoo Kim)

## Instructions:

1. Whenever possible, please provide output along with your code. For example, in Problem 2, you can print `url1` after defining it. Similarly, in Problem 6, print the first 5 rows of `df2` after creating it. Providing output as much as possible helps reduce the possibility of grading errors.
2. It is best practice to re-run everything before submitting it. This serves as a final check to ensure there are no errors in your code or output.


0. Import numpy and pandas.

In [1]:
import numpy as np
import pandas as pd

1. Import the `os` module. Print out the current working directory. 

In [2]:
import os
os.getcwd()

'/Users/ws/Projects/study_books/07_Programing_and_Data/Week_03'

2. Then create a variable, `url1` (your current working directory), and use `r` to capture the full url as a raw string. 

In [3]:
url1 = r'/Users/ws/Projects/study_books/07_Programing_and_Data/Week_03'
url1

'/Users/ws/Projects/study_books/07_Programing_and_Data/Week_03'

3. Create am empty folder `Financial Statement` under the current working directory using Python.

In [4]:
fs_directory = os.getcwd()+'/Financial Statement/'
if not os.path.exists(fs_directory):
    os.mkdir(fs_directory)
os.listdir(os.getcwd())

['.DS_Store',
 'MFIN290_Folders_and_Functions.ipynb',
 'Finished_Earnings',
 'MFIN290_Grouping_and_Summary_Stats.ipynb',
 'MFIN290_Missing_Data.ipynb',
 'MFIN290_Stacking_and_Merging.ipynb',
 'fina.csv',
 'Earnings.zip',
 'Earnings',
 'MFIN290_Reading_and_Filtering_Data.ipynb',
 'earnings1.csv',
 'Financial Statement',
 'MFIN290_HW3.ipynb']

4. Read the `fina.csv` file, and adjust the `datadate` column to the correct date format as we did in our class.

In [5]:
df = pd.read_csv('fina.csv', parse_dates=['datadate'], date_format='%Y%m%d')
df.head()

gvkey   datadate  fyr  tic      cik            conm       ni      sale  \
0   2575 2012-12-31   12  MTD  1037646  Mettler Toledo  290.847  2341.528   
1   2575 2013-12-31   12  MTD  1037646  Mettler Toledo  306.094  2378.972   
2   2575 2014-12-31   12  MTD  1037646  Mettler Toledo  338.241  2485.983   
3   2575 2015-12-31   12  MTD  1037646  Mettler Toledo  352.820  2395.447   
4   2575 2016-12-31   12  MTD  1037646  Mettler Toledo  384.370  2508.257   

      cogs    oancf        at        lt      seq  
0  811.204  327.704  2117.400  1290.181  827.219  
1  794.915  345.928  2152.819  1217.767  935.052  
2  809.537  418.912  2009.110  1289.515  719.595  
3  744.867  426.868  2018.485  1438.028  580.457  
4  767.753  443.078  2166.777  1731.834  434.943

5. Create a new DataFrame, `df2`, containing only the `gvkey` column.

In [6]:
df2 = df[['gvkey']]
df2.head()

gvkey
0   2575
1   2575
2   2575
3   2575
4   2575

6. Remove duplicate observations in `df2` to retain unique values in the `gvkey` column.

In [7]:
df2 = df2.drop_duplicates(ignore_index=True)
df2.head()

gvkey
0   2575
1   5977
2   6683
3   7334
4   9402

7. By utilizing `iterrows()` on `df2` and filtering `df` for each unique gvkey (e.g., df['gvkey'] == 2575), generate a subset DataFrame corresponding to each unique firm associated with that `gvkey` (e.g., df[df['gvkey'] == 2575]). Save each subset as a CSV file under the `Financial Statement` folder. For example, if `gvkey` is 2575, the file name should be `fina_2575.csv`. (Hint: To include an integer as a variable name, use code similar to `subset_df.to_csv(directory + f'fina_{gvkey}.csv', index=False)`)

In [8]:
for index, row in df2.iterrows():
    gvkey = row['gvkey']
    subset_df = df[df['gvkey']==gvkey]
    subset_df.to_csv(fs_directory + f'fina_{gvkey}.csv', index=False)

8. Create an empty list named `dfs`. Read each file in the `Financial Statement` folder with a subset of columns including `gvkey`, `datadate`, `at`, and `lt`, and append each DataFrame to the `dfs` list.

In [9]:
dfs = []
csv_list = os.listdir(fs_directory)
for csv in csv_list:
    subset_df = pd.read_csv(fs_directory+csv, index_col=None, header=0)
    subset_df = subset_df[['gvkey', 'datadate', 'at', 'lt']]
    dfs.append(subset_df)
dfs[0]

gvkey    datadate        at        lt
0  797235  2012-12-31  119213.0  100229.0
1  797235  2013-12-31  126223.0  103294.0
2  797235  2014-12-31  117532.0  105518.0
3  797235  2015-12-31  110495.0   96071.0
4  797235  2016-12-31  117470.0   99078.0
5  797235  2017-12-31  125356.0  107631.0
6  797235  2018-12-31  123382.0  106452.0
7  797235  2019-12-31  152186.0  131202.0

9. Concatenate all the data in `dfs` into a single DataFrame named `fina`.

In [10]:
fina = pd.concat(dfs, axis=0, ignore_index=True)
fina.head()

gvkey    datadate        at        lt
0  797235  2012-12-31  119213.0  100229.0
1  797235  2013-12-31  126223.0  103294.0
2  797235  2014-12-31  117532.0  105518.0
3  797235  2015-12-31  110495.0   96071.0
4  797235  2016-12-31  117470.0   99078.0

10. Calculate the median of `at` by `gvkey` using `groupby()`

In [11]:
fina.groupby('gvkey')['at'].median()

gvkey
2575        2159.798
5977      188602.000
6683        3451.048
7334       28316.500
9402      184958.500
             ...    
995990     11460.000
996343      7666.223
996812     23394.000
997531      8426.000
998876      1705.900
Name: at, Length: 481, dtype: float64